File: src/main/resources/harness/workspace/SandboxTest.js
/**
 * SandboxTest.js — the JavaScript twin of SandboxTest.java, and the contrast is the point.
 *
 * HOW TO SEE IT. The filter is ON by default -- `unsafe,java.io` -- so just run this. To watch the same
 * file with it off, restart the harness with `-Dcgui.harness.scriptPolicy=none` and compare.
 *
 * WHY THIS FILE READS DIFFERENTLY FROM THE JAVA ONE. Rhino enforces through a ClassShutter, which is
 * asked at the moment a script touches a class -- so JavaScript is refused ONE REACH AT A TIME, where
 * Java is refused as a whole file before it starts. That is not an inconsistency to be fixed; it is what
 * each engine can honestly do. Compiled Java links its classes when the class is defined, so the only
 * two places to stand are before it loads and inside the loader. Rhino resolves a name every time the
 * script asks for one, so it can answer every time.
 *
 * The practical difference is visible below: each section prints, then fails, then the next one runs.
 * Nothing above a refusal is un-done. In the Java file the refusal happens instead of the script.
 *
 * WHAT THIS IS NOT. Read plan_syntax.md 19.1. JavaScript's is the stronger of the two -- call-time
 * interception is real -- but the trust model is still "a script is code the player installed", and
 * server-authored scripts executing on a client are out of scope permanently. This is a guardrail.
 */

'use strict';

print('If you can read this, the script started -- JavaScript is refused per reach, not per file.');

function attempt(label, body) {
    try {
        print(label + ': ' + body());
    } catch (refused) {
        print(label + ' REFUSED: ' + refused);
    }
}

// ── 1. Ordinary I/O — refused by `java.io` in the example above ──────────────────────────────────
// `Java.type` is the front door, and it is the one a shutter sees first.
attempt('1. java.io.File', function () {
    var File = Java.type('java.io.File');
    return new File('.').getAbsolutePath();
});

// ── 2. The package-chain spelling of the same thing ──────────────────────────────────────────────
// `java.io.File` written out reaches the identical class by a different route. A filter that caught
// only `Java.type` would be a filter on a spelling rather than on a class.
attempt('2. package chain', function () {
    return 'reached java.io.File, isDirectory=' + new java.io.File('.').isDirectory();
});

// ── 3. Reflection — refused by `unsafe`, and the reason the rest is worth anything ───────────────
attempt('3. reflection', function () {
    // A STATIC METHOD, because a class object reached through Java.type answers its STATICS -- neither
    // `.name` nor `.class` is a member of it, and asking for either is an InternalError about this
    // fixture rather than a refusal by the filter. Both spellings were tried here and both misreported.
    var Modifier = Java.type('java.lang.reflect.Modifier');
    return 'reached java.lang.reflect.Modifier, isPublic(1)=' + Modifier.isPublic(1);
});

// ── 4. Leaving the JVM entirely — refused by `unsafe` ────────────────────────────────────────────
attempt('4. Runtime', function () {
    return Java.type('java.lang.Runtime').getRuntime().availableProcessors();
});

// ── 5. Something the policy ALLOWS, last, on purpose ─────────────────────────────────────────────
// A filter that refused everything would be indistinguishable from a broken engine. This line runs
// under the example policy above, which is how you tell the difference.
attempt('5. allowed', function () {
    var list = new java.util.ArrayList();
    list.add('still reachable');
    return list.get(0);
});

print('Done. Sections 1-4 refused, 5 ran -- that is the shutter working, not the engine failing.');

File: src/main/resources/harness/workspace/SandboxTest.java
/**
 * SandboxTest.java — a file written to be REFUSED. Every section reaches for something a locked-down
 * deployment would not allow.
 *
 * HOW TO SEE IT. The filter is ON by default -- `unsafe,java.io` -- so running this file should REFUSE
 * it outright: nothing below prints, and the Run panel shows which classes it reached for. To watch the
 * same file with the filter off, restart the harness with `-Dcgui.harness.scriptPolicy=none`.
 *
 * `unsafe` is ScriptPolicy.UNSAFE -- reflection, method handles, ClassLoader, Runtime, ProcessBuilder,
 * java.security and the internals. `java.io` is added so section 1 has something to hit.
 *
 * WHAT SHOULD HAPPEN, and the difference between the two halves is the whole design:
 *
 *   Sections 1-3 name their classes in source, so they are in the compiled constant pool. The
 *   AHEAD-OF-TIME scan (RefusedTypes) sees them and refuses the WHOLE FILE before a single line runs --
 *   which is why the first `println` below never appears. That matters: a script refused halfway has
 *   already done whatever it did before the refusal, and a partly-applied script is its own hazard.
 *
 *   Section 4 builds a class name at run time, so the constant pool holds only `java.lang.Class`. The
 *   scan cannot see it and passes; the LOADER GATE (ScriptClassLoader.loadClass) catches it when the
 *   name is actually resolved. To watch that half on its own, comment out sections 1-3.
 *
 * WHAT THIS IS NOT. Read plan/lang-stack.md 19.1. For Java this is a guardrail, not a security boundary:
 * compiled bytecode links what it links, SecurityManager is gone from modern JVMs, and a script that can
 * reach reflection can resolve names neither half of this ever sees. It stops accidents and casual
 * reach, and it keeps the editor from teaching an API the runtime will refuse. It does not contain a
 * determined author, and nothing here should be described as though it did.
 */

System.out.println("If you can read this, the class filter is OFF.");

// ── 1. Ordinary I/O — refused by `java.io` in the example above ──────────────────────────────────
// Nothing exotic. This is the shape most policies actually care about: a script reading or writing
// files on the machine of whoever is running it.
java.io.File here = new java.io.File(".");
System.out.println("1. listed " + here.getAbsolutePath());

// ── 2. Reflection — refused by `unsafe`, and the reason the rest is worth anything ───────────────
// The route out of every class filter. With this reachable, an allowlist is a suggestion: the name
// below is a string, and a string can be assembled.
java.lang.reflect.Method[] methods = String.class.getMethods();
System.out.println("2. String has " + methods.length + " methods");

// ── 3. Leaving the JVM entirely — refused by `unsafe` ────────────────────────────────────────────
// The one nobody argues about. `Runtime` and `ProcessBuilder` run whatever the host user can run.
Runtime runtime = Runtime.getRuntime();
System.out.println("3. " + runtime.availableProcessors() + " processors");

// ── 4. A name built at run time — the LOADER's case, not the scan's ──────────────────────────────
// The constant pool for this section names `java.lang.Class` and `java.lang.String` and nothing else,
// so the ahead-of-time scan has nothing to object to. Comment out 1-3 and run again: the file starts,
// prints, and is stopped at the moment the class is resolved rather than before it began.
String assembled = "java.io" + "." + "File";
try {
    Class<?> late = Class.forName(assembled);
    System.out.println("4. resolved " + late.getName() + " at run time");
} catch (ClassNotFoundException refused) {
    System.out.println("4. refused: " + refused.getMessage());
}
